pkg/dataobj/internal/dataset/reader_basic.go

```diff
@@ -168,6 +166,15 @@ func (pr *basicReader) fill(ctx context.Context, columns []Column, s []Row) (n i
 				return n, fmt.Errorf("column %v is not owned by basicReader", column)
 			}
 
+			// We want to allow readers to reuse memory of [Value]s in s while
+			// allowing the caller to retain ownership over that memory; to do this
+			// safely, we copy memory from s into pr.buf (for the given column index)
+			// for our decoders to use.
+			//
+			// If we didn't do this, then memory backing [Value]s are owned by both
+			// basicReader and the caller, which can lead to memory reuse bugs.
+			pr.buf = reuseRowsBuffer(pr.buf, s[n:], columnIndex)
```

Comment:
@cyriltovena @owen-d I found a future memory reuse bug that would happen. 

There's no way to trigger it now, but it's going to be a massive pain to track down once it does happen (e.g., once we switch over to `[]byte` instead of `string` and reuse the existing `[]byte` slice from a Value), so I fixed it in advance. 

(Similar code was added to `pageReader`, another place where the same bug could happen)

Reply:
I updated the tests to catch if this breaks in the future